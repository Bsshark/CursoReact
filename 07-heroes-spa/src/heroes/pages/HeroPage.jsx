import React from 'react'
import { Navigate, useNavigate, useParams } from 'react-router-dom'
import { getHeroById } from '../helpers';

export const HeroPage = () => {

  const {id} = useParams();

  const hero = getHeroById(id);

  const navigate = useNavigate();

  const onReturn = () => {
    navigate(-1);
  }

  if ( !hero ) {
    return <Navigate to="/marvel"></Navigate>
  }

  return (
    <div className='row mt-5'>
      <div className="col-4">
        <img 
        src={`/assets/heroes/${id}.jpg`} 
        alt={hero.superhero}
        className='img-thumbnail' />
      </div>
      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className='list-group list-group-flush'>
          <li className='list-group-item'><b>Alter ego:</b> { hero.alter_ego }</li>
          <li className='list-group-item'><b>Publisher:</b> {hero.publisher} </li>
          <li className='list-group-item'><b>First appareance:</b> {hero.first_appearance} </li>
        </ul>

        <h5 className='mt-3'>
          Characters
        </h5>
        <p className='mt-1'>
          {hero.characters}
        </p>

        <button 
          className='btn btn-primary mt-5'
          onClick={onReturn}>
            Regresar
        </button>
      </div>
    </div>
  )
}
