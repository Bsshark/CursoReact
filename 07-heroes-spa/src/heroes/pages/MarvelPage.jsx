import { HeroList } from "../components/heroList"

export const MarvelPage = () => {
  return (
    <>
    <h1>DC Comics</h1>
    <hr />

    <HeroList publisher={'Marvel Comics'}>

    </HeroList>

    </>
  )
}
