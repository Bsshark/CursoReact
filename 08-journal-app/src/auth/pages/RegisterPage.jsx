import { Alert, Button, Grid, Link, TextField, Typography } from "@mui/material"
import { Link as RouterLink } from 'react-router-dom';
import { AuthLayout } from "../layout/AuthLayout";
import { useForm } from "../../hooks";
import { useMemo, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { startCreatingUserWithEmailPassword } from "../../store/auth/thunks";

const formData = {
  email: '',
  password: '',
  displayName: ''
}

const formValidations = {
  email: [ (value) => value.includes('@'), 'El correo debe de tener una @'],
  password: [ (value) => value.length >= 6 , 'La password debe tener al menos 6 caracteres'],
  displayName: [ (value) => value.length >= 1, 'El nombre es obligatorio']
}

export const RegisterPage = () => {

  const dispatch = useDispatch();
  const [formSubmitted, setFormSubmitted] = useState(false);

  const {status, errorMessage} = useSelector(state => state.auth);
  const isCheckingauthentication = useMemo(() => status === 'checking', [status] );

  const { 
    formState, displayName, email, password, onInputChange,
    isFormValid, displayNameValid, emailValid, passwordValid
  } = useForm(formData, formValidations);


  const onSubmit = (event) => {
    event.preventDefault();
    setFormSubmitted(true);

    if(!isFormValid) return;

    dispatch(startCreatingUserWithEmailPassword(formState));
  }

  return (
    <AuthLayout title='Registro'>
      <h1>FormValid: {isFormValid ? 'Válido' : 'Incorrecto'}</h1>
      <form onSubmit={onSubmit} className="animate__animated animate__fadeIn animate__faster">
        <Grid container>
          <Grid item xs={12} sx={{mt: 2}}>
            <TextField 
              label="Nombre completo" 
              type="text" 
              placeholder="John Doe" 
              fullWidth
              name="displayName"
              value={displayName}
              onChange={onInputChange}
              error={ !!displayNameValid && formSubmitted}
              helperText= { displayNameValid }
              />

          </Grid>
          <Grid item xs={12} sx={{mt: 2}}>
            <TextField 
              label="Correo" 
              type="email" 
              placeholder="correo@gmail.com" 
              fullWidth
              name="email"
              error={!!emailValid && formSubmitted}
              helperText= { emailValid }
              value={email}
              onChange={onInputChange}
              />

          </Grid>
          <Grid item xs={12} sx={{mt: 2}}>
            <TextField 
              label="Contraseña" 
              type="password" 
              placeholder="contraseña" 
              fullWidth
              name="password"
              error={!!passwordValid && formSubmitted}
              value={password}
              helperText={passwordValid}
              onChange={onInputChange}/>

          </Grid>
          <Grid container spacing={ 2 } sx={{mb: 2, mt: 1}}>
          <Grid item xs={ 12 }  display={!!errorMessage ? '': 'none'}>
            <Alert severity="error">
                {errorMessage}
            </Alert>
            </Grid>
            <Grid item xs={ 12 } >
              <Button disabled={isCheckingauthentication} type="submit" variant="contained" fullWidth>Crear Cuenta</Button>
            </Grid>
          </Grid>

          <Grid container direction='row' justifyContent='end'>
            <Typography sx={{mr: 1}}>¿Ya tienes cuenta?</Typography>
            <Link component={ RouterLink } color='inherit' to="/auth/login">
              Ingresar
            </Link>

          </Grid>

        </Grid>

      </form>
    </AuthLayout>
  )
}
