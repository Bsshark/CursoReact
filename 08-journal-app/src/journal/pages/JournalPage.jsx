import { AddOutlined, MailOutline } from "@mui/icons-material"
import { IconButton, Typography } from "@mui/material"
import { JournalLayout } from "../layout/JournalLayout"
import { NoteView, NothingSelectedView } from "../views/"

export const JournalPage = () => {
  return (
    <JournalLayout>
      {/* <Typography>
        Dolor nisi quis sunt duis magna do proident proident irure sint esse.
      </Typography> */}
      <NothingSelectedView />
      {/* <NoteView/> */}

      <IconButton 
        size="large"
        sx={{
          color: 'white',
          bgcolor: 'error.main',
          ':hover': {
            bgcolor: 'error.main',
            opacity: 0.9
          },
          position: 'fixed',
          right: 50,
          bottom: 50
        }}
      >
        <AddOutlined sx={{fontSize: 30}}/>

      </IconButton>
    </JournalLayout>
    
  )
}
