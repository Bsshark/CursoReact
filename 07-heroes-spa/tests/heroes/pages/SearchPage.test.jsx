import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mockedUseNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUseNavigate
}))

describe('Pruebas en Search Page', () => { 

    beforeEach(() => jest.clearAllMocks());

    it('should show itself correctly with default values', () => {
        
        const {container} = render(
            <MemoryRouter>
                <SearchPage />
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
    });
    it('should show Batman and input with queryString value', () => {

        const searchValue = 'Batman';

        render(
            <MemoryRouter initialEntries={['/search?q=' + searchValue]}>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        expect(input.value).toBe(searchValue);
        const img = screen.getByRole('img');
        expect(img.src).toContain('dc-batman.jpg');

        const noHeroDiv = screen.getByTestId('alertNoHeroDiv');
        expect(noHeroDiv.style.display).toBe('none');
    });
    it('should show error if hero not found', () => {
        const searchValue = 'Batman123';

        render(
            <MemoryRouter initialEntries={['/search?q=' + searchValue]}>
                <SearchPage />
            </MemoryRouter>
        );

        const noHeroDiv = screen.getByTestId('alertNoHeroDiv');
        const heroDiv = screen.getByTestId('alertHeroDiv');
        expect(noHeroDiv.style.display).toBe('');
        expect(heroDiv.style.display).toBe('none');

    });

    it('should call navigate to the new screen (Batman123)', () => {
        
        const inputValue = 'Batman123';
        const onSubmit = jest.fn();

        render(
            <MemoryRouter initialEntries={['/search']}>
                <SearchPage />
            </MemoryRouter>
        );

        const input = screen.getByRole('textbox');
        fireEvent.change(input, {target: {value: inputValue}}); //nice

        const formSearch = screen.getByTestId('formSearch');
        fireEvent.submit(formSearch);

        expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${inputValue.toLowerCase().trim()}`);
        

        screen.debug();

    });

 })