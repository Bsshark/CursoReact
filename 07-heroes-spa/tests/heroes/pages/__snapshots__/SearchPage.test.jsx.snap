// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Search Page should show itself correctly with default values 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Searching
      </h4>
      <hr />
      <form
        data-testid="formSearch"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero name..."
          type="text"
          value=""
        />
        <button
          class="btn btn-outline-primary mt-1"
          data-testid="clickSubmitBtn"
        >
          Search
        </button>
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
        data-testid="alertHeroDiv"
      >
        Search a Hero
      </div>
      <div
        class="alert alert-danger animate__animated animate__fadeIn"
        data-testid="alertNoHeroDiv"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
